File: src/detection/yolov8_detector.py
"""
YOLOv8 目标检测器

支持:
  - ONNX Runtime 推理（跨平台）
  - A1 NPU 推理（通过 SmartSens SDK）
  - Letterbox 预处理，NMS 后处理
"""


import time
import numpy as np
from typing import List, Optional, Tuple
from dataclasses import dataclass, field

from src.utils.logger import logger

try:
    import cv2
except ImportError:
    cv2 = None

try:
    import onnxruntime as ort
    _ORT_AVAILABLE = True
except ImportError:
    _ORT_AVAILABLE = False


@dataclass
class Detection:
    """单个检测结果"""
    bbox: Tuple[int, int, int, int]  # 边界框：x1, y1, x2, y2
    confidence: float  # 置信度
    class_id: int  # 类别ID
    class_name: str = ""  # 类别名称

    @property
    def center(self) -> Tuple[int, int]:
        """获取边界框中心点坐标"""
        x1, y1, x2, y2 = self.bbox
        return ((x1 + x2) // 2, (y1 + y2) // 2)

    @property
    def area(self) -> int:
        """计算边界框面积"""
        x1, y1, x2, y2 = self.bbox
        return max(0, x2 - x1) * max(0, y2 - y1)

    @property
    def width_height(self) -> Tuple[int, int]:
        """获取边界框宽高"""
        x1, y1, x2, y2 = self.bbox
        return (x2 - x1, y2 - y1)



# COCO class names (80 classes)
COCO_CLASSES = [
    "person", "bicycle", "car", "motorcycle", "airplane", "bus", "train", "truck",
    "boat", "traffic light", "fire hydrant", "stop sign", "parking meter", "bench",
    "bird", "cat", "dog", "horse", "sheep", "cow", "elephant", "bear", "zebra",
    "giraffe", "backpack", "umbrella", "handbag", "tie", "suitcase", "frisbee",
    "skis", "snowboard", "sports ball", "kite", "baseball bat", "baseball glove",
    "skateboard", "surfboard", "tennis racket", "bottle", "wine glass", "cup",
    "fork", "knife", "spoon", "bowl", "banana", "apple", "sandwich", "orange",
    "broccoli", "carrot", "hot dog", "pizza", "donut", "cake", "chair", "couch",
    "potted plant", "bed", "dining table", "toilet", "tv", "laptop", "mouse",
    "remote", "keyboard", "cell phone", "microwave", "oven", "toaster", "sink",
    "refrigerator", "book", "clock", "vase", "scissors", "teddy bear",
    "hair drier", "toothbrush",
]


class YOLOv8Detector:
    """
    YOLOv8 推理引擎，支持 ONNX/NPU 后端
    """

    def __init__(self, config: dict):
        """初始化 YOLOv8 检测器"""
        self._cfg = config
        self._model_path = config.get("model_path", "models/yolov8n.onnx")  # 模型路径
        self._model_type = config.get("model_type", "onnx")  # 模型类型
        self._input_size = tuple(config.get("input_size", [320, 320]))  # 输入尺寸
        self._conf_thresh = config.get("conf_threshold", 0.45)  # 置信度阈值
        self._iou_thresh = config.get("iou_threshold", 0.5)  # IoU 阈值
        self._max_det = config.get("max_detections", 50)  # 最大检测数量
        self._classes = config.get("classes", None)  # 目标类别
        self._device = config.get("device", "cpu")  # 运行设备

        self._session = None  # 推理会话
        self._input_name = None  # 输入名称
        self._output_names = None  # 输出名称

        logger.info(
            f"YOLOv8Detector: model={self._model_path} "
            f"input={self._input_size} device={self._device}"
        )


    def initialize(self) -> bool:
        """加载模型并创建推理会话"""
        try:
            if self._model_type == "onnx":
                return self._init_onnx()
            elif self._model_type == "sdk_npu":
                return self._init_npu()
            else:
                logger.error(f"未知模型类型: {self._model_type}")
                return False
        except Exception as e:
            logger.error(f"模型初始化失败: {e}")
            return False


    def _init_onnx(self) -> bool:
        """初始化 ONNX Runtime 会话"""
        if not _ORT_AVAILABLE:
            logger.error("onnxruntime 未安装")
            return False

        providers = ["CPUExecutionProvider"]
        if self._device == "npu":
            # A1 NPU 可能需要自定义执行提供程序
            logger.info("请求 NPU，使用 ONNX 的 CPU 回退")

        self._session = ort.InferenceSession(self._model_path, providers=providers)
        self._input_name = self._session.get_inputs()[0].name
        self._output_names = [o.name for o in self._session.get_outputs()]

        logger.info(
            f"ONNX 模型已加载: input={self._input_name} "
            f"outputs={self._output_names}"
        )
        return True


    def _init_npu(self) -> bool:
        """初始化 A1 NPU 推理"""
        npu_cfg = self._cfg.get("npu_config", {})
        model_bin = npu_cfg.get("model_bin", "")
        model_param = npu_cfg.get("model_param", "")

        logger.info(f"初始化 NPU: bin={model_bin} param={model_param}")

        # SmartSens NPU API 占位符:
        # self._npu_net = smartsens_sdk.npu.load_model(model_bin, model_param)
        # self._npu_net.set_input_size(*self._input_size)

        logger.warning("NPU 初始化是占位符 - 使用 ONNX 回退")
        return self._init_onnx()


    # ── 推理 ─────────────────────────────────────────

    def detect(self, frame: np.ndarray) -> List[Detection]:
        """
        在帧上运行 YOLOv8 检测

        参数:
            frame: BGR 或灰度图像

        返回:
            Detection 对象列表
        """

        if self._session is None:
            logger.warning("Detector not initialized")
            return []

        t0 = time.time()

        # Preprocess
        input_tensor, ratio, (dw, dh) = self._preprocess(frame)

        # Inference
        outputs = self._session.run(self._output_names, {self._input_name: input_tensor})

        # Postprocess
        detections = self._postprocess(outputs, frame.shape, ratio, dw, dh)

        elapsed = (time.time() - t0) * 1000
        logger.debug(f"Detection: {len(detections)} objects in {elapsed:.1f}ms")

        return detections

    def _preprocess(self, frame: np.ndarray) -> Tuple[np.ndarray, float, Tuple[int, int]]:
        """
        Letterbox 调整大小 + 归一化

        返回:
            (input_tensor, ratio, (pad_w, pad_h))
        """

        if len(frame.shape) == 2:
            frame = cv2.cvtColor(frame, cv2.COLOR_GRAY2BGR)

        img_h, img_w = frame.shape[:2]
        inp_w, inp_h = self._input_size

        # Scale ratio
        ratio = min(inp_w / img_w, inp_h / img_h)
        new_w, new_h = int(img_w * ratio), int(img_h * ratio)
        dw, dh = (inp_w - new_w) // 2, (inp_h - new_h) // 2

        # Resize
        resized = cv2.resize(frame, (new_w, new_h), interpolation=cv2.INTER_LINEAR)

        # Pad
        padded = np.full((inp_h, inp_w, 3), 114, dtype=np.uint8)
        padded[dh:dh + new_h, dw:dw + new_w] = resized

        # Normalize: HWC → CHW, /255, float32, add batch dim
        blob = padded.astype(np.float32) / 255.0
        blob = blob.transpose(2, 0, 1)  # CHW
        blob = np.expand_dims(blob, axis=0)  # NCHW

        return blob, ratio, (dw, dh)

    def _postprocess(
        self,
        outputs: list,
        orig_shape: tuple,
        ratio: float,
        dw: int,
        dh: int,
    ) -> List[Detection]:
        """
        后处理 YOLOv8 输出: 解码边界框，应用 NMS

        YOLOv8 输出形状: [1, num_classes+4, num_anchors]
        """

        predictions = outputs[0]  # [1, 4+nc, N]

        # Transpose to [N, 4+nc]
        if predictions.shape[1] < predictions.shape[2]:
            predictions = predictions.transpose(0, 2, 1)  # [1, N, 4+nc]
        predictions = predictions[0]  # [N, 4+nc]

        # Extract boxes (cx, cy, w, h) and class scores
        boxes_cxcywh = predictions[:, :4]
        scores_all = predictions[:, 4:]

        # Get max class score and class id per box
        class_ids = np.argmax(scores_all, axis=1)
        confidences = scores_all[np.arange(len(class_ids)), class_ids]

        # Filter by confidence
        mask = confidences > self._conf_thresh
        if self._classes is not None:
            class_mask = np.isin(class_ids, self._classes)
            mask = mask & class_mask

        boxes_cxcywh = boxes_cxcywh[mask]
        confidences = confidences[mask]
        class_ids = class_ids[mask]

        if len(boxes_cxcywh) == 0:
            return []

        # Convert cx,cy,w,h → x1,y1,x2,y2
        boxes_xyxy = np.zeros_like(boxes_cxcywh)
        boxes_xyxy[:, 0] = boxes_cxcywh[:, 0] - boxes_cxcywh[:, 2] / 2  # x1
        boxes_xyxy[:, 1] = boxes_cxcywh[:, 1] - boxes_cxcywh[:, 3] / 2  # y1
        boxes_xyxy[:, 2] = boxes_cxcywh[:, 0] + boxes_cxcywh[:, 2] / 2  # x2
        boxes_xyxy[:, 3] = boxes_cxcywh[:, 1] + boxes_cxcywh[:, 3] / 2  # y2

        # Remove padding and rescale to original image
        boxes_xyxy[:, [0, 2]] = (boxes_xyxy[:, [0, 2]] - dw) / ratio
        boxes_xyxy[:, [1, 3]] = (boxes_xyxy[:, [1, 3]] - dh) / ratio

        # Clip to image bounds
        orig_h, orig_w = orig_shape[:2]
        boxes_xyxy[:, [0, 2]] = np.clip(boxes_xyxy[:, [0, 2]], 0, orig_w)
        boxes_xyxy[:, [1, 3]] = np.clip(boxes_xyxy[:, [1, 3]], 0, orig_h)

        # NMS
        indices = self._nms(boxes_xyxy, confidences, self._iou_thresh)
        indices = indices[:self._max_det]

        # Build Detection objects
        detections = []
        for i in indices:
            x1, y1, x2, y2 = boxes_xyxy[i].astype(int)
            cid = int(class_ids[i])
            cname = COCO_CLASSES[cid] if cid < len(COCO_CLASSES) else str(cid)
            detections.append(Detection(
                bbox=(x1, y1, x2, y2),
                confidence=float(confidences[i]),
                class_id=cid,
                class_name=cname,
            ))

        return detections

    @staticmethod
    def _nms(boxes: np.ndarray, scores: np.ndarray, iou_threshold: float) -> List[int]:
        """非极大值抑制 (纯 numpy 实现)"""

        if len(boxes) == 0:
            return []

        x1 = boxes[:, 0]
        y1 = boxes[:, 1]
        x2 = boxes[:, 2]
        y2 = boxes[:, 3]
        areas = (x2 - x1) * (y2 - y1)

        order = scores.argsort()[::-1]
        keep = []

        while len(order) > 0:
            i = order[0]
            keep.append(i)

            if len(order) == 1:
                break

            xx1 = np.maximum(x1[i], x1[order[1:]])
            yy1 = np.maximum(y1[i], y1[order[1:]])
            xx2 = np.minimum(x2[i], x2[order[1:]])
            yy2 = np.minimum(y2[i], y2[order[1:]])

            inter = np.maximum(0, xx2 - xx1) * np.maximum(0, yy2 - yy1)
            iou = inter / (areas[i] + areas[order[1:]] - inter + 1e-6)

            remain = np.where(iou <= iou_threshold)[0]
            order = order[remain + 1]

        return keep

    # ── Utils ─────────────────────────────────────────────

    def draw_detections(
        self, frame: np.ndarray, detections: List[Detection]
    ) -> np.ndarray:
        """在帧上绘制边界框和标签"""

        if cv2 is None:
            return frame

        vis = frame.copy()
        for det in detections:
            x1, y1, x2, y2 = det.bbox
            color = self._class_color(det.class_id)
            cv2.rectangle(vis, (x1, y1), (x2, y2), color, 2)
            label = f"{det.class_name} {det.confidence:.2f}"
            (tw, th), _ = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 1)
            cv2.rectangle(vis, (x1, y1 - th - 6), (x1 + tw, y1), color, -1)
            cv2.putText(
                vis, label, (x1, y1 - 4),
                cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1,
            )
        return vis

    @staticmethod
    def _class_color(class_id: int) -> Tuple[int, int, int]:
        """为类别生成一致的颜色"""

        np.random.seed(class_id + 42)
        return tuple(int(c) for c in np.random.randint(50, 255, 3))

    def release(self):
        """释放模型资源"""
        self._session = None
        logger.info("YOLOv8 检测器已释放")

File: src/src/core/yolov8_detector.cpp
#include "yolov8_detector.h"
#include "../utils/logger.h"
#include "../utils/exception_handler.h"

#include <algorithm>
#include <chrono>
#include <cstring>
#include <cmath>

// Optional: ONNX Runtime
#ifdef USE_ONNXRUNTIME
#include <onnxruntime_cxx_api.h>
#endif

static const char* MOD = "Detector";

// COCO 80 classes
static const char* COCO_CLASSES[] = {
    "person","bicycle","car","motorcycle","airplane","bus","train","truck",
    "boat","traffic light","fire hydrant","stop sign","parking meter","bench",
    "bird","cat","dog","horse","sheep","cow","elephant","bear","zebra",
    "giraffe","backpack","umbrella","handbag","tie","suitcase","frisbee",
    "skis","snowboard","sports ball","kite","baseball bat","baseball glove",
    "skateboard","surfboard","tennis racket","bottle","wine glass","cup",
    "fork","knife","spoon","bowl","banana","apple","sandwich","orange",
    "broccoli","carrot","hot dog","pizza","donut","cake","chair","couch",
    "potted plant","bed","dining table","toilet","tv","laptop","mouse",
    "remote","keyboard","cell phone","microwave","oven","toaster","sink",
    "refrigerator","book","clock","vase","scissors","teddy bear",
    "hair drier","toothbrush"
};
static const int NUM_COCO_CLASSES = 80;

struct YOLOv8Detector::Impl {
#ifdef USE_ONNXRUNTIME
    Ort::Env env{ORT_LOGGING_LEVEL_WARNING, "yolov8"};
    Ort::SessionOptions sessionOpts;
    std::unique_ptr<Ort::Session> session;
    std::string inputName;
    std::vector<std::string> outputNames;
    std::vector<int64_t> inputShape;
    int numClasses = 80;
#endif
    bool loaded = false;
};

YOLOv8Detector::YOLOv8Detector() : impl_(std::make_unique<Impl>()) {}

YOLOv8Detector::~YOLOv8Detector() { release(); }

bool YOLOv8Detector::initialize(const std::string& modelPath,
                                 int inputW, int inputH,
                                 float confThresh, float iouThresh,
                                 int maxDet, int numThreads) {
    inputW_ = inputW;
    inputH_ = inputH;
    confThresh_ = confThresh;
    iouThresh_ = iouThresh;
    maxDet_ = maxDet;

    inputBlob_.resize(3 * inputW * inputH);

    LOG_I(MOD, "Loading model: %s (%dx%d, conf=%.2f, iou=%.2f, threads=%d)",
          modelPath.c_str(), inputW, inputH, confThresh, iouThresh, numThreads);

#ifdef USE_ONNXRUNTIME
    try {
        impl_->sessionOpts.SetIntraOpNumThreads(numThreads);
        impl_->sessionOpts.SetGraphOptimizationLevel(
            GraphOptimizationLevel::ORT_ENABLE_ALL);

        impl_->session = std::make_unique<Ort::Session>(
            impl_->env, modelPath.c_str(), impl_->sessionOpts);

        Ort::AllocatorWithDefaultOptions alloc;

        // Input info
        auto inputNamePtr = impl_->session->GetInputNameAllocated(0, alloc);
        impl_->inputName = inputNamePtr.get();

        auto inputInfo = impl_->session->GetInputTypeInfo(0);
        auto tensorInfo = inputInfo.GetTensorTypeAndShapeInfo();
        impl_->inputShape = tensorInfo.GetShape();

        // Output info
        size_t numOutputs = impl_->session->GetOutputCount();
        for (size_t i = 0; i < numOutputs; ++i) {
            auto name = impl_->session->GetOutputNameAllocated(i, alloc);
            impl_->outputNames.push_back(name.get());
        }

        // Determine number of classes from output shape
        auto outInfo = impl_->session->GetOutputTypeInfo(0);
        auto outShape = outInfo.GetTensorTypeAndShapeInfo().GetShape();
        if (outShape.size() >= 2) {
            // [1, 4+nc, N] or [1, N, 4+nc]
            int dim1 = (int)outShape[1];
            int dim2 = outShape.size() > 2 ? (int)outShape[2] : 0;
            impl_->numClasses = (dim1 > dim2 ? dim1 : dim2) - 4;
            if (impl_->numClasses <= 0 || impl_->numClasses > 1000)
                impl_->numClasses = 80;
        }

        impl_->loaded = true;
        ready_ = true;
        LOG_I(MOD, "Model loaded: input=%s shape=[%ld,%ld,%ld,%ld] classes=%d",
              impl_->inputName.c_str(),
              impl_->inputShape.size() > 0 ? impl_->inputShape[0] : 0,
              impl_->inputShape.size() > 1 ? impl_->inputShape[1] : 0,
              impl_->inputShape.size() > 2 ? impl_->inputShape[2] : 0,
              impl_->inputShape.size() > 3 ? impl_->inputShape[3] : 0,
              impl_->numClasses);
        return true;

    } catch (const std::exception& e) {
        LOG_E(MOD, "ONNX Runtime init failed: %s", e.what());
        ExceptionHandler::instance().handle("OnnxInitError", e.what(), "detection");
        return false;
    }
#else
    LOG_W(MOD, "ONNX Runtime not available, detector will return empty results");
    ready_ = false;
    return false;
#endif
}

void YOLOv8Detector::preprocess(const uint8_t* frame, int width, int height,
                                 float* output, float& ratio,
                                 float& dw, float& dh) {
    // Letterbox resize from grayscale to RGB 320x320
    float scaleW = (float)inputW_ / width;
    float scaleH = (float)inputH_ / height;
    ratio = std::min(scaleW, scaleH);

    int newW = (int)(width * ratio);
    int newH = (int)(height * ratio);
    dw = (inputW_ - newW) / 2.0f;
    dh = (inputH_ - newH) / 2.0f;

    int idw = (int)dw;
    int idh = (int)dh;

    // Fill with 114/255 = 0.447
    int total = 3 * inputW_ * inputH_;
    float padVal = 114.0f / 255.0f;
    for (int i = 0; i < total; ++i) output[i] = padVal;

    // Resize with bilinear interpolation and convert GREY to 3-channel
    for (int y = 0; y < newH; ++y) {
        float srcY = (float)y / ratio;
        int sy0 = (int)srcY;
        int sy1 = std::min(sy0 + 1, height - 1);
        float fy = srcY - sy0;

        int dstY = y + idh;
        if (dstY >= inputH_) continue;

        for (int x = 0; x < newW; ++x) {
            float srcX = (float)x / ratio;
            int sx0 = (int)srcX;
            int sx1 = std::min(sx0 + 1, width - 1);
            float fx = srcX - sx0;

            // Bilinear interpolation
            float val = (1.0f - fy) * ((1.0f - fx) * frame[sy0 * width + sx0] +
                                        fx * frame[sy0 * width + sx1]) +
                        fy * ((1.0f - fx) * frame[sy1 * width + sx0] +
                              fx * frame[sy1 * width + sx1]);

            float normalized = val / 255.0f;
            int dstX = x + idw;
            if (dstX >= inputW_) continue;

            // NCHW format: 3 channels (grayscale replicated)
            int pixIdx = dstY * inputW_ + dstX;
            output[0 * inputW_ * inputH_ + pixIdx] = normalized;  // R
            output[1 * inputW_ * inputH_ + pixIdx] = normalized;  // G
            output[2 * inputW_ * inputH_ + pixIdx] = normalized;  // B
        }
    }
}

std::vector<Detection> YOLOv8Detector::detect(const uint8_t* frame,
                                                int width, int height) {
    if (!ready_) return {};

    auto t0 = std::chrono::steady_clock::now();

    float ratio, dw, dh;
    preprocess(frame, width, height, inputBlob_.data(), ratio, dw, dh);

    std::vector<Detection> results;

#ifdef USE_ONNXRUNTIME
    try {
        std::vector<int64_t> inputShape = {1, 3, inputH_, inputW_};

        Ort::MemoryInfo memInfo = Ort::MemoryInfo::CreateCpu(
            OrtArenaAllocator, OrtMemTypeDefault);

        Ort::Value inputTensor = Ort::Value::CreateTensor<float>(
            memInfo, inputBlob_.data(), inputBlob_.size(),
            inputShape.data(), inputShape.size());

        const char* inputNames[] = {impl_->inputName.c_str()};
        std::vector<const char*> outputNames;
        for (auto& n : impl_->outputNames) outputNames.push_back(n.c_str());

        auto outputs = impl_->session->Run(
            Ort::RunOptions{nullptr},
            inputNames, &inputTensor, 1,
            outputNames.data(), outputNames.size());

        // Parse output
        auto& outTensor = outputs[0];
        auto outShape = outTensor.GetTensorTypeAndShapeInfo().GetShape();
        const float* outData = outTensor.GetTensorData<float>();

        int dim1 = (int)outShape[1];
        int dim2 = outShape.size() > 2 ? (int)outShape[2] : 1;

        // YOLOv8 output: [1, 4+nc, N] -> transpose if needed
        int numAnchors, numValues;
        const float* predData = outData;
        std::vector<float> transposed;

        if (dim1 < dim2) {
            // [1, 4+nc, N] -> transpose to [N, 4+nc]
            numAnchors = dim2;
            numValues = dim1;
            transposed.resize(dim1 * dim2);
            for (int i = 0; i < dim1; ++i) {
                for (int j = 0; j < dim2; ++j) {
                    transposed[j * dim1 + i] = outData[i * dim2 + j];
                }
            }
            predData = transposed.data();
        } else {
            numAnchors = dim1;
            numValues = dim2;
        }

        int nc = numValues - 4;
        results = postprocess(predData, numAnchors, nc,
                              width, height, ratio, dw, dh);

    } catch (const std::exception& e) {
        LOG_E(MOD, "Inference error: %s", e.what());
        ExceptionHandler::instance().handleException(e, "detection");
    }
#endif

    auto t1 = std::chrono::steady_clock::now();
    lastInferMs_ = std::chrono::duration<float, std::milli>(t1 - t0).count();

    return results;
}

std::vector<Detection> YOLOv8Detector::postprocess(
    const float* data, int numAnchors, int numClasses,
    int origW, int origH, float ratio, float dw, float dh) {

    std::vector<Detection> dets;

    for (int i = 0; i < numAnchors; ++i) {
        const float* row = data + i * (4 + numClasses);

        float cx = row[0], cy = row[1], w = row[2], h = row[3];

        // Find best class
        int bestClass = 0;
        float bestScore = row[4];
        for (int c = 1; c < numClasses; ++c) {
            if (row[4 + c] > bestScore) {
                bestScore = row[4 + c];
                bestClass = c;
            }
        }

        if (bestScore < confThresh_) continue;

        // Convert cx,cy,w,h to x1,y1,x2,y2
        Detection det;
        det.x1 = (cx - w * 0.5f - dw) / ratio;
        det.y1 = (cy - h * 0.5f - dh) / ratio;
        det.x2 = (cx + w * 0.5f - dw) / ratio;
        det.y2 = (cy + h * 0.5f - dh) / ratio;

        // Clip
        det.x1 = std::max(0.0f, std::min(det.x1, (float)origW));
        det.y1 = std::max(0.0f, std::min(det.y1, (float)origH));
        det.x2 = std::max(0.0f, std::min(det.x2, (float)origW));
        det.y2 = std::max(0.0f, std::min(det.y2, (float)origH));

        det.confidence = bestScore;
        det.classId = bestClass;
        strncpy(det.className, className(bestClass), sizeof(det.className) - 1);

        dets.push_back(det);
    }

    // NMS
    auto keep = nms(dets, iouThresh_);
    std::vector<Detection> result;
    int limit = std::min((int)keep.size(), maxDet_);
    for (int i = 0; i < limit; ++i) {
        result.push_back(dets[keep[i]]);
    }

    return result;
}

float YOLOv8Detector::iou(const Detection& a, const Detection& b) {
    float xx1 = std::max(a.x1, b.x1);
    float yy1 = std::max(a.y1, b.y1);
    float xx2 = std::min(a.x2, b.x2);
    float yy2 = std::min(a.y2, b.y2);

    float inter = std::max(0.0f, xx2 - xx1) * std::max(0.0f, yy2 - yy1);
    float areaA = (a.x2 - a.x1) * (a.y2 - a.y1);
    float areaB = (b.x2 - b.x1) * (b.y2 - b.y1);

    return inter / (areaA + areaB - inter + 1e-6f);
}

std::vector<int> YOLOv8Detector::nms(const std::vector<Detection>& dets,
                                      float iouThreshold) {
    if (dets.empty()) return {};

    // Sort by confidence
    std::vector<int> indices(dets.size());
    for (int i = 0; i < (int)dets.size(); ++i) indices[i] = i;
    std::sort(indices.begin(), indices.end(),
              [&](int a, int b) { return dets[a].confidence > dets[b].confidence; });

    std::vector<int> keep;
    std::vector<bool> suppressed(dets.size(), false);

    for (int idx : indices) {
        if (suppressed[idx]) continue;
        keep.push_back(idx);

        for (int j : indices) {
            if (suppressed[j] || j == idx) continue;
            if (iou(dets[idx], dets[j]) > iouThreshold) {
                suppressed[j] = true;
            }
        }
    }

    return keep;
}

const char* YOLOv8Detector::className(int id) {
    if (id >= 0 && id < NUM_COCO_CLASSES) return COCO_CLASSES[id];
    return "unknown";
}

void YOLOv8Detector::release() {
#ifdef USE_ONNXRUNTIME
    if (impl_) {
        impl_->session.reset();
        impl_->loaded = false;
    }
#endif
    ready_ = false;
    LOG_I(MOD, "Detector released");
}